src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5875,304 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
+
+    if (m_row_delimiter.size())
+      csv.row_delimiter = *m_row_delimiter.c_str();
+
+    if (m_column_delimiter.size())
+      csv.column_delimiter = *m_column_delimiter.c_str();
+
+    if (m_quot.size())
+      csv.quot_char = *m_quot.c_str();
+
+    if (m_escape_char.size())
+      csv.escape_char = *m_escape_char.c_str();
+
+    if(m_header_info.compare("IGNORE")==0)
+    {
+      csv.ignore_header_info=true;
+    }
+    else if(m_header_info.compare("USE")==0)
+    {
+      csv.use_header_info=true;
+    } 
+    
+    m_s3_csv_object = new s3selectEngine::csv_object(s3select_syntax,csv);
+  }
+
+  if (s3select_syntax->get_error_description().empty() == false)
+  {
+    header_size = create_header_records(m_buff_header);
+
+    m_result.append(m_buff_header,header_size); 
+
+    m_result.append(PAYLOAD_LINE);
+    
+    m_result.append(s3select_syntax->get_error_description());
+
+    ldout(s->cct, 10) << "s3-select query: failed to prase query; {" << s3select_syntax->get_error_description() << "}"<< dendl;
+
+    status = -1;
+  }
+  else
+  {
+    header_size = create_header_records(m_buff_header);
+
+    m_result.append(m_buff_header,header_size); 
+
+    m_result.append(PAYLOAD_LINE);
+
+    status = m_s3_csv_object->run_s3select_on_stream(m_result,input,input_length,s->obj_size);
+    if(status<0)
+    {
+      m_result.append(m_s3_csv_object->get_error_description());
+    }
+  }
+
+  if (m_result.size() > strlen(PAYLOAD_LINE))
+  {
+    m_result.append(END_PAYLOAD_LINE);
+
+    int buff_len = create_message(m_result.data(), m_result.size() - 12, header_size);
+
+    s->formatter->write_bin_data(m_result.data(), buff_len);
+    if (op_ret < 0)
+      return op_ret;
+  }
+  rgw_flush_formatter_and_reset(s, s->formatter);
+
+  return status;
+}
+
+int RGWSelectObj_ObjStore_S3::handle_aws_cli_parameters(std::string & sql_query)
+{
+
+  if(chunk_number !=0)
+    return 0;
+
+#define GT "&gt;"
+#define LT "&lt;"
+  if (m_s3select_query.find(GT) != std::string::npos)
+    boost::replace_all(m_s3select_query, GT, ">");
+  if (m_s3select_query.find(LT) != std::string::npos)
+    boost::replace_all(m_s3select_query, LT, "<");
+
+  //AWS cli s3select parameters
+  extract_by_tag("Expression", sql_query);
+
+  extract_by_tag("FieldDelimiter", m_column_delimiter);
+
+  extract_by_tag("QuoteCharacter", m_quot);
+
+  extract_by_tag("RecordDelimiter", m_row_delimiter);
+  if (m_row_delimiter.size()==0) 
+    m_row_delimiter='\n';
+
+  extract_by_tag("QuoteEscapeCharacter", m_escape_char);
+
+  extract_by_tag("CompressionType", m_compression_type);
+
+  if (m_compression_type.length()>0 && m_compression_type.compare("NONE") != 0)
+  {
+      ldout(s->cct, 10) << "RGW supports currently only NONE option for compression type" << dendl;
+      return -1;
+  }
+
+  extract_by_tag("FileHeaderInfo",m_header_info);
+
+  return 0;
+}
+
+int RGWSelectObj_ObjStore_S3::extract_by_tag(std::string tag_name,std::string & result)
```

Comment:
can tag_name be passed by reference as well?  I'm guessing not, because we extracted it from something?

Reply:
'extract_by_tag'  is called with rvalue only as tag_name (above), 
this method is called few-times per query.